src/cargo/core/compiler/job_queue.rs

```diff
@@ -167,6 +166,12 @@ pub struct JobState<'a> {
     /// Channel back to the main thread to coordinate messages and such.
     messages: Arc<Queue<Message>>,
 
+    /// Normally output is sent to the job queue with backpressure. When the job is fresh
+    /// however we need to immediately display the output to prevent a deadlock as the
+    /// output messages are processed on the same thread as they are sent from. `output`
+    /// defines where to output in this case.
+    output: Option<&'a Config>,
```

Comment:
Can you include some notes that it is crucial that this is only set on the main thread?  There are several concerns:

* The `RefMut` from shell can cause panics if used between threads.
* The event loop is carefully crafted to avoid flickering of the progress bar.  By splitting the output into different places, this makes it harder to ensure batches of messages are grouped together. I think the change is OK because it is all on the main thread, but it is pretty subtle.
* Keeping one thread responsible for output helps prevent interleaving of messages. In particular, some messages are not printed atomically (like the "status" messages).

Reply:
The `Config` is not `Sync`, so it isn't possible to assign it when not running on the main thread. Also because `Shell` is wrapped in a `RefCell` it isn't possible to borrow it twice at the same time to interleave output.

Reply:
I see.  Still, please add a comment discussing the concerns about the coordination of ownership of the output.  My concern about interleaving is about the future, and things to watch out for if this is ever changed (like wrapping output in a mutex).  It might also help to explain how `JobState` works with respect to how it is constructed and passed into the job threads.

Reply:
Interleaving would still not be possible when the `RefCell` is replaced with a `Mutex` as the mutex would be locked for the duration of the printing of a single message. It would be hard to accidentally unlock the mutex in the middle of printing.

I can add some more docs to `JobState`.

Reply:
Done